src/foam/nanos/mrac/ClusterDAO.js

```diff
@@ -0,0 +1,182 @@
+    },
+    {
+      documentation: `Array of all clients to put/remove to after a non-server operation.`,
+      name: 'clients',
+      class: 'FObjectArray',
+      of: 'foam.dao.DAO',
+      visibility: 'HIDDEN'
+    },
+    {
+      documentation: `Reference to the ClusterConfigDAO to which we are subscribed/listening for updates to reconfigure.`,
+      name: 'clusterConfigDAO',
+      class: 'foam.dao.DAOProperty',
+      visibility: 'HIDDEN'
+    }
+  ],
+
+  methods: [
+    {
+      documentation: `Upon initialization create the ClusterServer configuration and register nSpec.`,
+      name: 'init_',
+      javaCode: `
+        reconfigure(getX());
+
+        /* register ClusterConfig Listener */
+        DAO clusterConfigDAO = (DAO) getX().get("clusterConfigDAO");
+        clusterConfigDAO.listen(new ClusterConfigSink(getX(), this), TRUE);
+        /* REVIEW: need to keep a reference to the dao?*/
+        setClusterConfigDAO(clusterConfigDAO);
+      `
+    },
+    {
+      documentation: `Rebuild the client list.`,
+      name: 'reconfigure',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        }
+      ],
+      javaCode: `
+      Logger logger = (Logger) x.get("logger");
+      logger.debug(this.getClass().getSimpleName(), "reconfigure", getNSpec().getName());
+      DAO dao = (DAO) x.get("clusterConfigDAO");
+      List arr = (ArrayList) ((ArraySink) dao
+       .where(
+         AND(
+           EQ(ClusterConfig.ENABLED, true),
+           EQ(ClusterConfig.STATUS, Status.ONLINE)
+         )
+       )
+       .select(new ArraySink())).getArray();
+      DAO[] newClients = new DAO[arr.size()];
+      for ( int i = 0; i < arr.size(); i++ ) {
+        ClusterConfig config = (ClusterConfig) arr.get(i);
+        if ( "localhost".equals(config.getId()) ) {
+          setConfig(config);
+          continue;
+        }
+        DAO client = new ClusterClientDAO.Builder(x).setServiceName(getNSpec().getName()).setConfig(config).build();
+        newClients[i] = client;
+      }
+      setClients(newClients);
+      `
+    },
+    {
+      name: 'put_',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'obj',
+          type: 'foam.core.FObject'
+        }
+      ],
+      javaCode: `
+      foam.core.FObject o = getDelegate().put_(x, obj);
+      for ( DAO client : getClients() ) {
+        client.put(o);
+      }
+      return o;
+     `
+    },
+    {
+      name: 'remove_',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'obj',
+          type: 'foam.core.FObject'
+        }
+      ],
+      javaCode: `
+      foam.core.FObject o = getDelegate().remove_(x, obj);
+      for ( DAO client : getClients() ) {
+        client.put(o);
+      }
+      return o;
+     `
+    },
+    {
+      name: 'cmd_',
+      javaCode: `
+      if ( obj instanceof ClusterRequest ) {
```

Comment:
Then finally we can re-write this method to unpack RPCMessages and apply them to the delegate.  You can look at foam/java/Skeleton.js for how we do this normally.

Reply:
A future improvement would be to add a method to RPCMessage called ".apply" which takes an object and does the method dispatch for us.